### PR TITLE
PERF: Datetimelike lookups

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @TomAugspurger this should address at least some of the regressions you posted about this morning.

### CODE REVIEW

File: pandas/core/arrays/_mixins.py [hunk 1]
```diff
@@ -64,22 +65,24 @@ def _validate_fill_value(self, fill_value):
...
+    #  the _index_data kludge in libreduction
     @property
     def shape(self) -> Tuple[int, ...]:
         return self._ndarray.shape
 
     def __len__(self) -> int:
         return self.shape[0]
 
-    @property
+    @cache_readonly
```
Comment (jreback) [inline pandas/core/arrays/_mixins.py:77]: really? these are not worth it i think

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -526,6 +528,13 @@ def __getitem__(self, key):
         only handle list-likes, slices, and integer scalars
         """
 
+        if lib.is_integer(key):
+            # fast-path
+            result = self._data[key]
+            if self.ndim == 1:
+                return self._box_func(result)
+            return self._simple_new(result, dtype=self.dtype)
+
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:537]: elif

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
does this link to a specific issue?

**Comment by jbrockmendel** [comment]
> does this link to a specific issue?

Tom opened a handful of perf-regression issues yesterday.  I havent checked which of those this affects, but I think the fastpath in `__getitem__` is a big one.

File: pandas/core/arrays/_mixins.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:77]: in profiling it made a bigger difference than i expected

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:537]: i think i prefer it this way.  the if/elif/else cases that follow this fit together nicely, dont return-early

**Review by jreback — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**